drivers/pinctrl/pinctrl-bflb.c

```diff
@@ -0,0 +1,802 @@
+{
+	struct bflb_gpio_pinctrl *pctl = gpiochip_get_data(chip);
+	unsigned int reg = bflb_gpio_get_reg(pctl, offset);
+
+	if (FIELD_GET(REG_GPIOx_OE, reg) == 1 &&
+			FIELD_GET(REG_GPIOx_IE, reg) == 0) {
+		return GPIO_LINE_DIRECTION_OUT;
+	} else if (FIELD_GET(REG_GPIOx_IE, reg) == 1 &&
+			FIELD_GET(REG_GPIOx_OE, reg) == 0) {
+		return GPIO_LINE_DIRECTION_IN;
+	}
+
+	return -EIO;
+}
+
+//AH: Get the incoming input or outgoing output value for the specified GPIO
+static int bflb_gpio_get(struct gpio_chip *chip, unsigned int offset)
+{
+	struct bflb_gpio_pinctrl *pctl = gpiochip_get_data(chip);
+	unsigned int reg = bflb_gpio_get_reg(pctl, offset);
+
+	if (FIELD_GET(REG_GPIOx_OE, reg) == 1 &&
+			FIELD_GET(REG_GPIOx_IE, reg) == 0) {
+		reg = readl_relaxed(pctl->base + REG_GPIO(offset));
+		return !!(reg & REG_GPIOx_O);
+	} else if (FIELD_GET(REG_GPIOx_IE, reg) == 1 &&
+			FIELD_GET(REG_GPIOx_OE, reg) == 0) {
+		reg = readl_relaxed(pctl->base + REG_GPIO(offset));
+		return !!(reg & REG_GPIOx_I);
+	}
+
+	return -EIO;
+}
+
+//AH: Set the specified GPIO's output state
+static void bflb_gpio_set(struct gpio_chip *chip, unsigned int offset,
+		int value)
+{
+	struct bflb_gpio_pinctrl *pctl = gpiochip_get_data(chip);
+
+	bflb_gpio_set_reg(pctl, offset, REG_GPIOx_O, value ?
+			FIELD_PREP(REG_GPIOx_O, 1) : 0);
+
+	dev_dbg(pctl->dev, "Pin %u set to value %u", offset, value);
+}
+
+//AH: Set the specified gpio direction to input
+static int bflb_gpio_direction_input(struct gpio_chip *chip,
+		unsigned int offset)
+{
+	struct bflb_gpio_pinctrl *pctl = gpiochip_get_data(chip);
+
+	bflb_gpio_set_reg
+	(
+		pctl,
+		offset,
+		REG_GPIOx_OE | REG_GPIOx_IE | REG_GPIOx_SMT,
+		FIELD_PREP(REG_GPIOx_OE, 0) | FIELD_PREP(REG_GPIOx_IE, 1) |
+			FIELD_PREP(REG_GPIOx_SMT, 1)
+	);
+
+	dev_dbg(pctl->dev, "Pin %u set to direction input", offset);
+
+	return 0;
+}
+
+//AH: Set the specified gpio direction to output
+static int bflb_gpio_direction_output(struct gpio_chip *chip,
+		unsigned int offset, int value)
+{
+	struct bflb_gpio_pinctrl *pctl = gpiochip_get_data(chip);
+
+	bflb_gpio_set_reg
+	(
+		pctl,
+		offset,
+		REG_GPIOx_PD | REG_GPIOx_PU | REG_GPIOx_OE | REG_GPIOx_IE |
+			REG_GPIOx_SMT | REG_GPIOx_MODE,
+		FIELD_PREP(REG_GPIOx_PD, 0) | FIELD_PREP(REG_GPIOx_PU, 0) |
+			FIELD_PREP(REG_GPIOx_OE, 1) | FIELD_PREP(REG_GPIOx_IE, 0) |
+			FIELD_PREP(REG_GPIOx_SMT, 1) | FIELD_PREP(REG_GPIOx_MODE, 0)
+	);
+
+	dev_dbg(pctl->dev, "Pin %u set to direction output", offset);
+
+	bflb_gpio_set(chip, offset, value); //Set the initially passed value
+
+	return 0;
+}
+
+//AH: Configure pin electrical characteristics
+static int bflb_gpio_set_config(struct gpio_chip *chip, unsigned int offset,
+		unsigned long config)
+{
+	struct bflb_gpio_pinctrl *pctl = gpiochip_get_data(chip);
+	enum pin_config_param param = pinconf_to_config_param(config);
+	unsigned int arg = pinconf_to_config_argument(config);
+
+	switch (param) {
+	case PIN_CONFIG_BIAS_DISABLE:
+		bflb_gpio_set_reg
```

Comment:
nit: LKML style encourages functions to fit on one screen, so arguments on separate lines is discouraged.